src/StateService/Storage/StateStore.cs

```diff
@@ -0,0 +1,174 @@
+        public uint LocalRootIndex => currentSnapshot.CurrentLocalRootIndex();
+        public uint ValidatedRootIndex => currentSnapshot.CurrentValidatedRootIndex();
+
+        private static StateStore singleton;
+        public static StateStore Singleton
+        {
+            get
+            {
+                while (singleton is null) Thread.Sleep(10);
+                return singleton;
+            }
+        }
+
+        public StateStore(NeoSystem core, string path)
+        {
+            if (singleton != null) throw new InvalidOperationException(nameof(StateStore));
+            this.core = core;
+            this.store = core.LoadStore(path);
+            singleton = this;
+            core.ActorSystem.EventStream.Subscribe(Self, typeof(Blockchain.RelayResult));
+            UpdateCurrentSnapshot();
+        }
+
+        public void Dispose()
+        {
+            store.Dispose();
+        }
+
+        public StateSnapshot GetSnapshot()
+        {
+            return new StateSnapshot(store);
+        }
+
+        public HashSet<byte[]> GetProof(UInt256 root, StorageKey skey)
+        {
+            using ISnapshot snapshot = store.GetSnapshot();
+            var trie = new MPTTrie<StorageKey, StorageItem>(snapshot, root);
+            return trie.GetProof(skey);
+        }
+
+        protected override void OnReceive(object message)
+        {
+            switch (message)
+            {
+                case StorageChanges changes:
+                    UpdateLocalStateRoot(changes.Height, changes.ChangeSet);
+                    break;
+                case Blockchain.RelayResult rr:
+                    if (rr.Result == VerifyResult.Succeed && rr.Inventory is ExtensiblePayload payload && payload.Category == StatePlugin.StatePayloadCategory)
+                        OnStatePayload(payload);
+                    break;
+                default:
+                    break;
+            }
+        }
+
+        private void OnStatePayload(ExtensiblePayload payload)
+        {
+            StateRoot state_root = null;
+            try
+            {
+                state_root = payload.Data?.AsSerializable<StateRoot>();
+            }
+            catch (Exception ex)
+            {
+                Utility.Log(nameof(StateStore), LogLevel.Warning, " invalid state root " + ex.Message);
+                return;
+            }
+            if (state_root != null)
+                OnNewStateRoot(state_root);
+        }
+
+        private bool OnNewStateRoot(StateRoot state_root)
+        {
+            if (state_root?.Witness is null) return false;
+            if (state_root.Index <= ValidatedRootIndex) return false;
+            if (LocalRootIndex < state_root.Index && state_root.Index < LocalRootIndex + MaxCacheCount)
+            {
+                cache.Add(state_root.Index, state_root);
+                return true;
+            }
+            using var state_snapshot = Singleton.GetSnapshot();
```

Comment:
I think thtat we should not dispose this Snapshot, it will be disposed when call `UpdateCurrentSnapshot`

Reply:
It won't. `GetSnapshot()` returns a new instance.

Reply:
> I think thtat we should not dispose this Snapshot, it will be disposed when call `UpdateCurrentSnapshot`

I don't think so.